File: bindings/go/satox/api/graphql_api.go
package api

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// GraphQLClient represents a GraphQL API client
type GraphQLClient struct {
	baseURL    string
	httpClient *http.Client
	apiKey     string
}

// GraphQLRequest represents a GraphQL request
type GraphQLRequest struct {
	Query         string                 `json:"query"`
	Variables     map[string]interface{} `json:"variables,omitempty"`
	OperationName string                 `json:"operationName,omitempty"`
}

// GraphQLResponse represents a GraphQL response
type GraphQLResponse struct {
	Data       json.RawMessage        `json:"data"`
	Errors     []GraphQLError         `json:"errors,omitempty"`
	Extensions map[string]interface{} `json:"extensions,omitempty"`
}

// GraphQLError represents a GraphQL error
type GraphQLError struct {
	Message    string                 `json:"message"`
	Locations  []ErrorLocation        `json:"locations,omitempty"`
	Path       []interface{}          `json:"path,omitempty"`
	Extensions map[string]interface{} `json:"extensions,omitempty"`
}

// ErrorLocation represents the location of a GraphQL error
type ErrorLocation struct {
	Line   int `json:"line"`
	Column int `json:"column"`
}

// NewGraphQLClient creates a new GraphQL client
func NewGraphQLClient(baseURL string, apiKey string) *GraphQLClient {
	return &GraphQLClient{
		baseURL: baseURL,
		httpClient: &http.Client{
			Timeout: time.Second * 30,
		},
		apiKey: apiKey,
	}
}

// Execute sends a GraphQL request and returns the response
func (c *GraphQLClient) Execute(ctx context.Context, req GraphQLRequest) (*GraphQLResponse, error) {
	jsonBody, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.baseURL, bytes.NewReader(jsonBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Set headers
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("X-API-Key", c.apiKey)

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	var graphqlResp GraphQLResponse
	if err := json.Unmarshal(body, &graphqlResp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	if len(graphqlResp.Errors) > 0 {
		return &graphqlResp, fmt.Errorf("graphql errors: %v", graphqlResp.Errors)
	}

	return &graphqlResp, nil
}

// Query sends a GraphQL query
func (c *GraphQLClient) Query(ctx context.Context, query string, variables map[string]interface{}, operationName string) (*GraphQLResponse, error) {
	return c.Execute(ctx, GraphQLRequest{
		Query:         query,
		Variables:     variables,
		OperationName: operationName,
	})
}

// Mutation sends a GraphQL mutation
func (c *GraphQLClient) Mutation(ctx context.Context, mutation string, variables map[string]interface{}, operationName string) (*GraphQLResponse, error) {
	return c.Execute(ctx, GraphQLRequest{
		Query:         mutation,
		Variables:     variables,
		OperationName: operationName,
	})
}

// SetTimeout sets the client timeout
func (c *GraphQLClient) SetTimeout(timeout time.Duration) {
	c.httpClient.Timeout = timeout
}

// SetAPIKey sets the API key
func (c *GraphQLClient) SetAPIKey(apiKey string) {
	c.apiKey = apiKey
}

// SetBaseURL sets the base URL
func (c *GraphQLClient) SetBaseURL(baseURL string) {
	c.baseURL = baseURL
}

// Example usage:
/*
client := NewGraphQLClient("https://api.example.com/graphql", "your-api-key")

query := `
query GetUser($id: ID!) {
    user(id: $id) {
        id
        name
        email
    }
}
`

variables := map[string]interface{}{
    "id": "123",
}

resp, err := client.Query(context.Background(), query, variables, "GetUser")
if err != nil {
    log.Fatal(err)
}

var result struct {
    User struct {
        ID    string `json:"id"`
        Name  string `json:"name"`
        Email string `json:"email"`
    } `json:"user"`
}

if err := json.Unmarshal(resp.Data, &result); err != nil {
    log.Fatal(err)
}

fmt.Printf("User: %+v\n", result.User)
*/

File: bindings/python/satox/api/graphql_api.py
"""
GraphQL API client for Satox.
"""

import json
from typing import Dict, Any, Optional, List
import requests
from requests.exceptions import RequestException

class GraphQLError(Exception):
    """Base class for GraphQL errors."""
    def __init__(self, message: str, locations: Optional[List[Dict[str, int]]] = None, path: Optional[List[str]] = None):
        self.message = message
        self.locations = locations
        self.path = path
        super().__init__(f"GraphQL Error: {message}")

class GraphQLResponse:
    """Represents a GraphQL response."""
    def __init__(self, data: Dict[str, Any], errors: Optional[List[Dict[str, Any]]] = None):
        self.data = data
        self.errors = errors
        if errors:
            raise GraphQLError(
                message=errors[0].get('message', 'Unknown error'),
                locations=errors[0].get('locations'),
                path=errors[0].get('path')
            )

class GraphQLClient:
    """GraphQL client for Satox."""
    
    def __init__(self, endpoint: str, api_key: str):
        self.endpoint = endpoint
        self.api_key = api_key
        self.session = requests.Session()
        self.session.headers.update(self._get_headers())

    def _get_headers(self) -> Dict[str, str]:
        """Get request headers."""
        return {
            'Authorization': f'Bearer {self.api_key}',
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        }

    def execute(self, query: str, variables: Optional[Dict[str, Any]] = None) -> GraphQLResponse:
        """Execute a GraphQL query."""
        payload = {
            'query': query,
            'variables': variables or {}
        }

        try:
            response = self.session.post(self.endpoint, json=payload)
            response.raise_for_status()
            result = response.json()
            return GraphQLResponse(
                data=result.get('data', {}),
                errors=result.get('errors')
            )
        except RequestException as e:
            raise GraphQLError(f"Request failed: {str(e)}")
        except json.JSONDecodeError as e:
            raise GraphQLError(f"Invalid response format: {str(e)}")

    def execute_mutation(self, mutation: str, variables: Optional[Dict[str, Any]] = None) -> GraphQLResponse:
        """Execute a GraphQL mutation."""
        return self.execute(mutation, variables)

    def execute_subscription(self, subscription: str, variables: Optional[Dict[str, Any]] = None) -> GraphQLResponse:
        """Execute a GraphQL subscription."""
        return self.execute(subscription, variables)

    def close(self):
        """Close the session."""
        self.session.close() 